event-sourcing/event-sourcing-core/src/main/scala/org/apache/james/eventsourcing/CommandDispatcher.scala

```diff
@@ -0,0 +1,82 @@
+/** **************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ * *
+ * http://www.apache.org/licenses/LICENSE-2.0                   *
+ * *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ * ***************************************************************/
+package org.apache.james.eventsourcing
+
+import com.google.common.base.Preconditions
+import javax.inject.Inject
+import org.apache.james.eventsourcing.eventstore.EventStoreFailedException
+
+import scala.util.{Failure, Success, Try}
+
+object CommandDispatcher {
+  private val MAX_RETRY = 10
+
+  case class UnknownCommandException(command: Command) extends RuntimeException(String.format("Unknown command %s", command)) {
+    def getCommand: Command = command
+  }
+
+  case class TooManyRetries(command: Command, retries: Int) extends RuntimeException(String.format("Too much retries for command %s. Store failure after %d retries", command, retries)) {
+    def getCommand: Command = command
+
+    def getRetries: Int = retries
+  }
+}
+
+class CommandDispatcher @Inject() (eventBus: EventBus, handlers: Set[CommandHandler[_ <: Command]]) {
+  Preconditions.checkArgument(handlers.groupBy(_.handledClass).values.forall(_.size == 1), "It should exist only one handler by command".asInstanceOf[Object])
+
+  private val handlersByClass: Map[Class[_ <: Command], CommandHandler[_ <: Command]] =
+    handlers.map(handler => (handler.handledClass, handler)).toMap
+
+  def dispatch(c: Command): Unit = {
+    trySeveralTimes(() => tryDispatch(c))
+      .getOrElse(() => throw CommandDispatcher.TooManyRetries(c, CommandDispatcher.MAX_RETRY))
+  }
+
+  private def trySeveralTimes(singleTry: () => Boolean): Option[Unit] =
+    0.until(CommandDispatcher.MAX_RETRY)
+      .find(_ => singleTry())
+      .map(_ => ())
+
+
+  private def tryDispatch(c: Command): Boolean = {
+      val maybeEvents: Option[Try[List[_ <: Event]]] = handleCommand(c)
+      maybeEvents match {
+        case Some(eventsTry) =>
+          eventsTry
```

Comment:
It will be treated in the pattern matching which apply on the resulting `Try`  (applying handler then publishing the events ) :
```
              case Failure(_ : EventStoreFailedException) => false
              case Failure(e) => throw e
```